### PR TITLE
Add new case for fwcfg64 basic testing

### PR DESCRIPTION
**Pull request by peixiu**
Fwcfg64 basic function test:
1) Boot guest with -device vmcoreinfo.
2) Check the fwcfgg64 driver has been installed.
3) Run "dump-guest-memory -w memory.dmp" in qemu monitor.
4) Check the memory.dmp can be saved and the size is larger then 0Kb.
5) Check the dump file can be open with windb tools.

id: 1992917
Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
PASS 1-Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.fwcfg64.q35

**Comment by peixiu** [comment]
Hi @leidwang @vivianQizhu

Could you help to review this patch?  This patch is based on https://github.com/avocado-framework/avocado-vt/pull/3199 and https://github.com/avocado-framework/avocado-vt/pull/3200, https://gitlab.cee.redhat.com/kvm-qe/kar/-/merge_requests/736

Thanks a lot~
Peixiu

File: provider/win_dump_utils.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    dump_name = utils_misc.generate_random_string(4) + "Memory.dmp"
+    dump_file = tmp_dir + "/" + dump_name
+
+    output = vm.monitor.human_monitor_cmd('dump-guest-memory -w %s'
+                                          % dump_file)
+    if output:
+        test.fail("Save dump file failed as: %s" % output)
+    else:
+        cmd = "ls -l %s | awk '{print $5}'" % dump_file
+        dump_size = int(process.run(cmd, shell=True).stdout_text)
```
Comment (vivianQizhu) [inline provider/win_dump_utils.py:32]: process.getoutput()

File: provider/win_dump_utils.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    Check Windows Debug Tools is installed
+
+    :param test: kvm test object
+    :param params: the dict used for parameters.
+    :param session: The guest session object.
+    """
+    check_cmd = params["chk_windbg_cmd"] % params["windbg_path"]
+    status, output = session.cmd_status_output(check_cmd)
+    if status:
+        test.fail("windbg tool has not been installed")
```
Comment (vivianQizhu) [inline provider/win_dump_utils.py:66]:
```suggestion
    if not utils_misc.wait_for(lambda: check_windbg_installed(params, session), timeout=timeout):
        test.fail("windbg tool has not been installed")


def check_windbg_installed(params, session):
    """
    Check Windows Debug Tools is installed

    :param test: kvm test object
    :param params: the dict used for parameters.
    :param session: The guest session object.
    """
    check_cmd = params["chk_windbg_cmd"] % params["windbg_path"]
    status, _ = session.cmd_status_output(check_cmd)
    return False if status else True
```

File: qemu/tests/fwcfg64.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    4) Check the memory.dmp can be saved and the size is larger then 0Kb.
+    5) Check the dump file can be open with windb tools.
+    """
+    def set_vm_param():
+        """
+        Update vm mem and image size params to match fwcfg64 test needed.
+
+        :param params: Params object
+        """
+        host_free_mem = int(process.run(params["get_free_mem"],
```
Comment (vivianQizhu) [inline qemu/tests/fwcfg64.py:27]: `utils_misc.get_free_mem`

File: qemu/tests/fwcfg64.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    Fwcfg64 basic function test:
+
+    1) Boot guest with -device vmcoreinfo.
+    2) Check the fwcfgg64 driver has been installed.
+    3) Run "dump-guest-memory -w memory.dmp" in qemu monitor.
+    4) Check the memory.dmp can be saved and the size is larger then 0Kb.
+    5) Check the dump file can be open with windb tools.
+    """
+    def set_vm_param():
```
Comment (vivianQizhu) [inline qemu/tests/fwcfg64.py:21]: Will this function be used by other fwcfg64 test cases? If so, I suppose it also could be added to win_dump_utils.py.

File: qemu/tests/fwcfg64.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        Update vm mem and image size params to match fwcfg64 test needed.
+
+        :param params: Params object
+        """
+        host_free_mem = int(process.run(params["get_free_mem"],
+                                        shell=True).stdout_text)
+        host_avail_disk = int(process.run(params["get_avail_disk"],
+                                          shell=True).stdout_text)
+        if host_avail_disk < (host_free_mem / 1024) + 30:
```
Comment (vivianQizhu) [inline qemu/tests/fwcfg64.py:31]: Would you help explain why `+30`? If it is guest mem, better get it from `params` since it could be changed.

Comment (vivianQizhu) [inline qemu/tests/fwcfg64.py:31]: Would you help explain why `+30`? If it is guest mem, better get it from default `params['mem']` as it is possibly to be configured.

File: qemu/tests/fwcfg64.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                                                  disk,
+                                                  params['image_size_stg'],
+                                                  params["os_type"])[0]
+    error_context.context("Generate the Memory.dmp file", logging.info)
+    dump_file = win_dump_utils.generate_mem_dump(test, params, vm)
+
+    error_context.context("Copy the Memory.dmp file"
+                          "from host to guest", logging.info)
+    dump_path = disk_letter
+    vm.copy_files_to(dump_file, "%s:\\Memory.dmp" % dump_path,
```
Comment (vivianQizhu) [inline qemu/tests/fwcfg64.py:57]:
```suggestion
    vm.copy_files_to(dump_file, "%s:\\Memory.dmp" % disk_letter,
```

File: qemu/tests/fwcfg64.py (hunk 5 above)
Comment (peixiu) [reply qemu/tests/fwcfg64.py:31]: 30 is for system image size, we need make sure the "host available disk size"  is larger than the "free memory + system image size", make sure host have  enough size to create a data disk image which size is same as the vm mem and can save a dump file which size same as vm mem as well.
 If not, we'll changed the vm mem calculation formula, based on 80 percent of the  "host available disk size - vm system image size",  vm_mem = (host_avail_disk - 30) * 0.8 // 2

Comment (peixiu) [reply qemu/tests/fwcfg64.py:31]: I'll try to update it with params["image_size"]

File: qemu/tests/fwcfg64.py (hunk 4 above)
Comment (peixiu) [reply qemu/tests/fwcfg64.py:21]: I'll move this function to win_dump_utils.py, thanks~

**Comment by vivianQizhu** [comment]
@peixiu Please check and fix the CI failure, and we'd better run inspekt check locally before submit code to CI.

**Comment by peixiu** [comment]
@vivianQizhu   I've updated all as your advice,  for get_free_mem() function, it does not support for host mem, I used get_mem_info() replace it, thank you a lot~

Tested with latest code:
PASS 1-Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.fwcfgdmp.q35

**Comment by leidwang** [comment]
05:23:18 ERROR| ERROR 2-Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.fwcfgdmp.q35 -> FileTransferTimeoutError: Timeout expired while sending data to server    (filename: /home/tmp/QzdxMemory.dmp)

It seems that the dmp file is too large and the transfer is not completed within the timeout time.Maybe we need adjust the timeout or limit the memory and dump file size.

**Comment by peixiu** [comment]
Tested with memory size '30720'
 (1/2) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.fwcfg.q35: PASS (777.63 s)
 (2/2) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.1.io-github-autotest-qemu.fwcfg.q35: PASS (773.73 s)

@leidwang @vivianQizhu Please help to review it again when you free, thanks a lot~

**Comment by leidwang** [comment]
23:52:30 INFO | PASS 1-Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win11.x86_64.io-github-autotest-qemu.fwcfg.q35

LGTM.

File: qemu/tests/fwcfg.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                                                                vm)
+
+    try:
+        error_context.context("Copy the Memory.dmp.zip file "
+                              "from host to guest", logging.info)
+        unzip_speed = int(params.get("unzip_speed", 80))
+        vm.copy_files_to(dump_zip_file, "%s:\\Memory.dmp.zip" % disk_letter)
+        unzip_cmd = params["unzip_cmd"] % (disk_letter, disk_letter)
+        unzip_timeout = int(params["mem"]) // unzip_speed
+        status, output = session.cmd_status_output(unzip_cmd, timeout=unzip_timeout)
```
Comment (vivianQizhu) [inline qemu/tests/fwcfg.py:55]: This line is too long.

File: qemu/tests/cfg/fwcfg.cfg [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    tmp_dir = "/var/tmp"
+    get_avail_disk = df -h /home | grep home | awk -F "G" '{print $3}'
+    unzip_cmd = 'powershell -command "Expand-Archive %s:\Memory.dmp.zip %s:"'
+    unzip_speed = 60
+    move_cmd = "move {0}:\*Memory.dmp {0}:\Memory.dmp"
+    chk_id_cmd = 'type %s | find /i "LIVE_SYSTEM_DUMP (161)"'
+    images += " stg"
+    image_name_stg = "images/storage"
+    remove_image_stg = yes
+    no ppc64 ppc64le
```
Comment (vivianQizhu) [inline qemu/tests/cfg/fwcfg.cfg:17]: What about aarch64?

Comment (peixiu) [reply qemu/tests/cfg/fwcfg.cfg:17]: Hi @vivianQizhu  Confirm with feature owner before, aarch64 is support vmcoreinfo device, only power pc does not support now~

Comment (vivianQizhu) [reply qemu/tests/cfg/fwcfg.cfg:17]: @peixiu So this test case works well on `aarch64` platform? Have you ever tested it or do we need to request them to help review and test?

Comment (peixiu) [reply qemu/tests/cfg/fwcfg.cfg]: @vivianQizhu Updated as our discussed, thanks~

File: provider/win_dump_utils.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    session.cmd(windbg_install_cmd)
+    if not utils_misc.wait_for(lambda: check_windbg_installed(params, session),
+                               timeout=timeout):
+        test.fail("windbg tool has not been installed")
+
+
+def check_windbg_installed(params, session):
+    """
+    Check Windows Debug Tools is installed
+    :param test: kvm test object
```
Comment (vivianQizhu) [inline provider/win_dump_utils.py:86]: There is no param `test` for this function, also please refine the docstring for all other functions in this module. Thanks.

**Review by vivianQizhu — COMMENTED** [review]
LGTM otherwise.

File: provider/win_dump_utils.py (hunk 9 above)
Comment (peixiu) [reply provider/win_dump_utils.py]: @vivianQizhu Updated, thanks~

**Comment by peixiu** [comment]
Tested with the latest codes:
(1/1) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.1.io-github-autotest-qemu.fwcfg.q35: PASS

**Pull request merged**